doc/design/thin_pool.md

```diff
@@ -0,0 +1,111 @@
+
++	// OverProvisionRatio represents the ratio of overprovision that can
++	// be allowed on thin pools
++	OverprovisionRatio int `json:"overprovisionRatio,omitempty"`
+}
+
+type DeviceClass struct {
+    Name string `json:"name,omitempty"`
+
+    DeviceSelector *DeviceSelector `json:"deviceSelector,omitempty"`
+    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+
++   // ThinPoolConfig contains configurations for the thin-pool
++   ThinPoolConfig  ThinPoolConfig `json:"thinPoolConfig,omitempty"`
+}
+```
+
+
+- Following new fields will added to `DeviceClass` API
+    - **ThinPoolConfig** API contains information related to a thin pool.These configuration options are:
+        - **Name**: Name of the thin-pool
+        - **SizePercent**: Size of the thin pool to be created with respect to available free space in the volume group. It represents percentage value and not absolute size values. Size value should range between 10-90. It defaults to 75 if no value is provided.
+        - **OverprovisionRatio**: OverprovisionRatio represents the ratio of overprovision that can be allowed on thin pools.
+
+- `LVMVolumeGroup` API changes:
+
+``` go=
+type LVMVolumeGroupSpec struct {
+    // DeviceSelector is a set of rules that should match for a device to be
+    // included in this TopoLVMCluster
+	// +optional
+    DeviceSelector *DeviceSelector `json:"deviceSelector,omitempty"`
+
+    // NodeSelector chooses nodes
+    // +optional
+    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+
++   // ThinPoolConfig contains configurations for the thin-pool
++   ThinPoolConfig ThinPoolConfig `json:"thinPoolConfig,omitempty"`
+}
+```
+
+### VolumeGroup Manager
+- Volume Group Manager is responsible for creating the thin-pools after creating the volume group.
+- Command used for creating the thin-pool:
+    ```
+    lvcreate -L <Size>%FREE -c <Chunk size> -T <vg_name>/<thin-pool-name>
+    ```
+    where:
+    - Size is `LVMClusterSpec.Storage.DeviceClass.ThinPoolConfig.SizePercent`
+    - chunk size is 512KiB, which is the default.
+
+- Manager should also update the `lvmd.yaml` file to map volume group and its thin-pool
+- Sample `lvmd.yaml` config file
+``` yaml=
+device-classes:
+  - name: ssd-thin
+    volume-group: myvg1
+    spare-gb: 10
+    type: thin
+    thin-pool-config:
+        name: pool0
+        overprovision-ratio: 50.0
+```
+
+### Monitoring and Alerts
+- Available thin-pool size (both data and metadata) should be provided by topolvm as prometheus metrics.
+- Threshold limits for the thin-pool should be provide as static values in the prometheus yaml file.
+- If available size of data or metadata for a particular thin-pool crosses the threshold, then appropriate alerts should be triggered.
+
+
+### Open questions
+- What should be the chunk size of the thin-pools?
+    - Use default size a 512 kiB for now.
+- Should LVMO notify changes of thin-pool if it happens to topolvm?
```

Comment:
Please clarify this point. What changes are you referring to ?

Reply:
- this question can be dropped, topolvm by default updates annotations in 10 min intervals if no ops performed in that time space